content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+![](https://i.imgur.com/4CDrXhL.png)
+
+Using PH, a viable block layout would be *M-A-B-C*, which we expand to *M-A0-A1-A2-B-C*. If we add up the weights of consecutive blocks (hereafter referred to as control flow transfers), we arrive at 180 (100 for *M-A0* and 80 for *A0-A1*). This is far from optimal, as we can generate the layout *M-A0-A1-B-A2-C* with 280 control flow transfers, an improvement of over 50% from before. Note that we use this metric because it tells us how often code will "fall-through." We want to maximize fall-throughs as this improves spatial locality.
+
+Another drawback of PH is that it uses an undirected call graph. As a result, in the eyes of PH, function *A* calling *B* is the same as *B* calling *A*, and PH treats ordering *A* before *B* as the same as ordering *B* before *A*. However, this is clearly not the case. Let's say *A* calls *B* 100 times, and *B* calls *A* 0 times. Then, if we order *A* before *B*, assuming the call occurs in the middle of *A*, the instruction pointer will have to jump *len(A)/2* instructions to the top of *B*, where *len(A)* denotes the number of instructions in *A*. Alternatively, if *B* was placed before *A*, then *A* would on average jump *len(B) + len(A)/2* instructions. Because of this discrepancy, it is important to take into account the caller and callee relations in the call graph.
+
+### Call-chain clustering
+[Call-chain clustering (C3)](https://research.fb.com/publications/optimizing-function-placement-for-large-scale-data-center-applications-2/) is a recent function layout optimization developed at Facebook Research that addresses the issue with using undirected call graphs. We do not discuss its internal details in this post, but a simplified way of thinking about C3 is that it is PH but with a directed call graph. In this way, it improves spatial locality by considering caller-callee relations.
+
+## Codestitcher
+
+This brings us to Codestitcher, an optimizer that integrates basic block reordering and function reordering to improve upon the PH model. Below is a high-level overview of the steps taken in Codestitcher to generate the optimized code layout, given an input program.
+
+1. Compile the program to LLVM bitcode. This serves as the baseline program. As the program is built, symbols are emitted that allow us to generate a CFG for the program.
+2. Profile the program, generating weights for CFG edges.
+3. Using the weighted CFG, perform basic block chaining, ordering basic blocks in a way to minimize jumps.
+4. Perform hierarchical code collocation, ordering these basic block chains to form higher-level layouts. Here, the *d*-close partial layout algorithm works to build layouts to maximize spatial locality.
+
+There is not much to say about step 1, so let us explore steps 2, 3, and 4 in more detail.
```

Comment:
TBH, I think this way they summarized the implementation was a little bogus. Indeed, who cares about compiling to LLVM? And I would even add, who cares about the step where you profile the CFG edges? _Obviously_ you have to do this to make any system like this work.

I think I would just say that this approach is going to work starting with an edge-profiled program, i.e., a CFG annotated with frequency counts on every edge. And that you can get that with profiling.

Then the only two steps that are actually part of Codestitcher are 3 and 4, the interesting ones.